lib/usb/Makefile

```diff
@@ -1,32 +1,59 @@
-USB_SRCS := \
-	$(NXDK_DIR)/lib/usb/host/ohci-hcd.c \
-	$(NXDK_DIR)/lib/usb/core/message.c \
-	$(NXDK_DIR)/lib/usb/core/hcd.c \
-	$(NXDK_DIR)/lib/usb/core/hcd-pci.c \
-	$(NXDK_DIR)/lib/usb/core/hub.c \
-	$(NXDK_DIR)/lib/usb/core/usb.c \
-	$(NXDK_DIR)/lib/usb/core/config.c \
-	$(NXDK_DIR)/lib/usb/core/urb.c \
-	$(NXDK_DIR)/lib/usb/core/buffer_simple.c \
-	$(NXDK_DIR)/lib/usb/core/usb-debug.c \
-	$(NXDK_DIR)/lib/usb/sys/BootUSB.c \
-	$(NXDK_DIR)/lib/usb/sys/linuxwrapper.c \
-	$(NXDK_DIR)/lib/usb/sys/xpad.c \
-	$(NXDK_DIR)/lib/usb/sys/xremote.c \
-	$(NXDK_DIR)/lib/usb/sys/usbkey.c \
-	$(NXDK_DIR)/lib/usb/sys/usbmouse.c \
-	$(NXDK_DIR)/lib/usb/misc/misc.c \
-	$(NXDK_DIR)/lib/usb/misc/pci.c \
-	$(NXDK_DIR)/lib/usb/misc/malloc.c
-
-USB_OBJS = $(addsuffix .obj, $(basename $(USB_SRCS)))
-
-$(NXDK_DIR)/lib/libnxdk_usb.lib: $(USB_OBJS)
-
-main.exe: $(NXDK_DIR)/lib/libnxdk_usb.lib
-
-CLEANRULES += clean-usb
-
-.PHONY: clean-usb
-clean-usb:
-	$(VE)rm -f $(USB_OBJS) $(NXDK_DIR)/lib/libnxdk_usb.lib
+USBH_DIR=$(NXDK_DIR)/lib/usb/ohci_usbhostlib
+
+#USB Host core
+USB_CORE_SRCS := \
+	$(USBH_DIR)/src_core/hub.c \
+	$(USBH_DIR)/src_core/mem_alloc.c \
+	$(USBH_DIR)/src_core/ohci.c \
+	$(USBH_DIR)/src_core/support.c \
+	$(USBH_DIR)/src_core/usb_core.c
+
+#USB Human Input Device class driver
+USB_HID_SRCS := \
+	$(USBH_DIR)/src_hid/hid_core.c \
+	$(USBH_DIR)/src_hid/hid_driver.c
+
+#USB Communication Device class driver
+USB_CDC_SRCS := \
+	$(USBH_DIR)/src_cdc/cdc_core.c \
+	$(USBH_DIR)/src_cdc/cdc_driver.c \
+	$(USBH_DIR)/src_cdc/cdc_parser.c
+	
+#USB Audio class driver
+USB_UAC_SRCS := \
+	$(USBH_DIR)/src_uac/uac_core.c \
+	$(USBH_DIR)/src_uac/uac_driver.c \
+	$(USBH_DIR)/src_uac/uac_parser.c
+
+# USB Video class driver
+USB_UVC_SRCS := \
+	$(USBH_DIR)/src_uvc/uvc_core.c \
+	$(USBH_DIR)/src_uvc/uvc_driver.c \
+	$(USBH_DIR)/src_uvc/uvc_parser.c
+
+# Mass Storage Class driver
+# FIXME: This driver is currently hardcoded so that it requires FATFS
+# This requirement should be removed and abstracted out.
+# USB_MSC_SRCS := \
+#	$(USBH_DIR)/src_msc/msc_driver.c \
+#	$(USBH_DIR)/src_msc/msc_xfer.c

```

Comment:
- What would happen if you enabled this right now? Does it auto-format devices or something?
- If this was more modular (nxdk_usb_msc.lib) then we could already build this, but disable default inclusion in default.xbe

Reply:
It has some things in the driver such as specific header includes, `#include "ff.h"`, and calls to fatfs specific functions like `f_mount`, 
This is specific to http://elm-chan.org/fsw/ff/00index_e.html. So the only way to compile it, is to include the whole fatfs lib. Ive done this here for testing purposes https://github.com/Ryzee119/nxdk_ohci_test.

Reply:
Okay. I didn't understand that from the existing comment (although it's quite clear, I thought that it requires use of the "FatFs" filesystem type, not some libraries).

 I think we should standardize an API so that the msc driver calls elsewhere.
Because USB drivers can already be out-of-tree, it would make sense to register filesystem handling out-of-tree as well.
However, I didn't look at that code yet.

That sounds like an upstreamable solution, too.

Temporarily disabling msc is fine with me.